src/analysis/retail/shaman/elemental/modules/talents/Stormkeeper.tsx

```diff
@@ -1,37 +1,264 @@
+
+    this.stSpender = this.selectedCombatant.hasTalent(TALENTS.ELEMENTAL_BLAST_TALENT)
+      ? TALENTS.ELEMENTAL_BLAST_TALENT
+      : TALENTS.EARTH_SHOCK_TALENT;
     this.active =
       this.selectedCombatant.getRepeatedTalentCount(TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT) > 0;
     if (!this.active) {
       return;
     }
+
+    this.addEventListener(Events.cast.by(SELECTED_PLAYER), this.onEventDuringSK);
+    this.addEventListener(
+      Events.cast.by(SELECTED_PLAYER).spell(TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT),
+      this.onSKCast,
+    );
     this.addEventListener(
-      Events.damage.by(SELECTED_PLAYER).spell(AFFECTED_ABILITIES),
+      Events.removebuff.by(SELECTED_PLAYER).spell(TALENTS.STORMKEEPER_1_ELEMENTAL_TALENT),
+      this.onSKFalloff,
+    );
+    this.addEventListener(
+      Events.damage.by(SELECTED_PLAYER).spell(SK_DAMAGE_AFFECTED_ABILITIES),
       this.onSKDamage,
     );
   }
 
+  onSKCast(cast: CastEvent) {
+    this.lastSKHardcast = cast;
+  }
+
+  onSKFalloff(event: RemoveBuffEvent) {
+    if (!this.activeWindow) {
+      return;
+    }
+
+    this.activeWindow.timeline.end = event.timestamp;
+
+    this.recordCooldown(this.activeWindow);
+    this.lastSKHardcast = null;
+    this.activeWindow = null;
+  }
+
+  onEventDuringSK(event: CastEvent) {
+    /* Don't include events that did not happen between hardcasts and the
+    SK buff falling off. */
+    if (!this.lastSKHardcast) {
+      return;
+    }
+
+    if (!this.activeWindow && WINDOW_START_SPELLS.includes(event.ability.guid)) {
+      this.activeWindow = {
+        event: this.lastSKHardcast,
+        maelstromOnCast: this.maelstromTracker.current,
+        flameshockDurationOnCast: this.enemies.getLongestDurationRemaining(
+          SPELLS.FLAME_SHOCK.id,
+          FLAMESHOCK_BASE_DURATION,
+          FLAMESHOCK_BASE_DURATION * (1 + PANDEMIC_WINDOW),
+          event.timestamp,
+        ),
+        electrifiedShocksDurationOnCast: this.enemies.getLongestDurationRemaining(
+          SPELLS.ELECTRIFIED_SHOCKS_DEBUFF.id,
+          ELECTRIFIED_SHOCKS_DURATION,
+          ELECTRIFIED_SHOCKS_DURATION,
+          event.timestamp,
+        ),
+        sopOnCast: this.selectedCombatant.hasBuff(SPELLS.SURGE_OF_POWER_BUFF.id),
+        moteOnCast: this.selectedCombatant.hasBuff(SPELLS.MASTER_OF_THE_ELEMENTS_BUFF.id),
+        timeline: {
+          start: this.lastSKHardcast.timestamp,
+          end: -1,
+          events: [this.lastSKHardcast],
+          performance: QualitativePerformance.Perfect,
+        },
+        firstRotationCast: event,
+      };
+
+      /* If the user started the window with a spender, then the maelstrom
+      has already been consumed */
+      if (event.ability.guid === this.stSpender.id) {
+        const resourceCost = event.resourceCost ? event.resourceCost : {};
+        this.activeWindow.maelstromOnCast += resourceCost[RESOURCE_TYPES.MAELSTROM.id] || 0;
+      }
+    }
+
+    if (this.activeWindow) {
```

Comment:
[Do] if you switch the order of these two if-statements I think you can get a better flow of "if this is an active window do this, otherwise if it is the start of the window then start the window".  It just means you can get rid of the `!this.activeWindow` above.

Reply:
I originally did as you suggested, but then I discovered that it needs to be in the original order. 

In your suggestion, the first spell of the window won't be processed by the code that determines if the player cast the spells in the correct order. For example if the first ability is an EB, then the code won't check and correctly mark if MotE was present before window start.

Do you feel I should refactor the code to account for this, or is the original order fine considering this use-case?